gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp

```diff
@@ -41,6 +41,31 @@
 
 #include <pcl/gpu/segmentation/gpu_extract_clusters.h>
 
+#include <cuda_runtime_api.h>
+#include <cuda.h>
+
+namespace pcl {
+namespace detail {
+
+// Downloads only the neccssary cluster indices from the device to the host.
+void economical_download(const pcl::gpu::NeighborIndices& source_indices,
```

Comment:
Shall we make it inline (or move the impl to cpp or make it templated)? It is a non-template function and I don't feel comfortable letting it go inside a header (multiple declarations are a possibility by inclusion in 2 cpp files which are linked together)

Reply:
This is an excellent question! While I understand your concerns, I am not used to working on such large codebases and have no experience in making an informed judgment. Do you have a recommendation?

Reply:
@larshg thoughts?

Reply:
Based on a quick skimming of https://stackoverflow.com/questions/1932311/when-to-use-inline-function-and-when-not-to-use-it

I would go for moving to a .cpp file.

- Its not a small function
- Its I/O bound (GPU?)

Reply:
Thank you very much, @larshg  for your comments! Do you know in which cpp file you would place the function? I see that [`extract_clusters.cpp`](https://github.com/PointCloudLibrary/pcl/blob/master/gpu/segmentation/src/extract_clusters.cpp) exists, but this file does not include [`gpu_extract_clusters.h`](https://github.com/PointCloudLibrary/pcl/blob/master/gpu/segmentation/include/pcl/gpu/segmentation/gpu_extract_clusters.h) yet and has no implementations in it.  

The function `economical_download` is indeed IO bound, spending almost all its time on device-host interaction.

Reply:
Just include gpu_extract_clusters.h and add the implementation. Seems to be have done something similar here: https://github.com/PointCloudLibrary/pcl/blob/master/recognition/src/cg/hough_3d.cpp

Reply:
great - thanks @larshg! I will do exactly that.

Reply:
PS: The impl hpp file should be including the h file for correct behavior, and including the h file in cpp file wouldn't be need if that's the case

Reply:
Thank you both, @larshg and @kunaltyagi,  for your detailed and helpful instructions! I placed the function `economical_download` in a source file and declared it in the header file. The function is a static member as I thought this is the most fitting namespace. Is this in line with your ideas?